src/main/storm/mesos/MesosNimbus.java

```diff
@@ -114,94 +120,229 @@ public String getHostName(Map<String, SupervisorDetails> map, String nodeId) {
+    final Protos.Filters filters = Protos.Filters.newBuilder()
+        .setRefuseSeconds(filterSeconds.intValue())
+        .build();
+    _offers = new RotatingMap<>(
+        new RotatingMap.ExpiredCallback<Protos.OfferID, Protos.Offer>() {
+          @Override
+          public void expire(Protos.OfferID key, Protos.Offer val) {
+            driver.declineOffer(
+                val.getId(),
+                filters
+            );
+          }
+        }
+    );
+
+    Number lruCacheSize = Optional.fromNullable((Number) _conf.get(CONF_MESOS_OFFER_LRU_CACHE_SIZE)).or(1000);
+    final int intLruCacheSize = lruCacheSize.intValue();
+    _usedOffers = Collections.synchronizedMap(new LinkedHashMap<Protos.TaskID, Protos.Offer>(intLruCacheSize + 1, .75F, true) {
+      // This method is called just after a new entry has been added
+      public boolean removeEldestEntry(Map.Entry eldest) {
+        return size() > intLruCacheSize;
+      }
+    });
+
+    Number offerExpired = Optional.fromNullable((Number) _conf.get(Config.NIMBUS_MONITOR_FREQ_SECS)).or(60);
+    Number expiryMultiplier = Optional.fromNullable((Number) _conf.get(CONF_MESOS_OFFER_EXPIRY_MULTIPLIER)).or(2.5);
+    _timer.scheduleAtFixedRate(new TimerTask() {
+      @Override
+      public void run() {
+        try {
+          synchronized (_offersLock) {
+            _offers.rotate();
+          }
+        } catch (Throwable t) {
+          LOG.error("Received fatal error Halting process...", t);
+          Runtime.getRuntime().halt(2);
+        }
+      }
+    }, 0, Math.round(1000 * expiryMultiplier.doubleValue() * offerExpired.intValue()));
+  }
+
+  public void shutdown() throws Exception {
+    _httpServer.shutdown();
+  }
+
+  public void resourceOffers(SchedulerDriver driver, List<Protos.Offer> offers) {
+    synchronized (_offersLock) {
+      if (_offers == null) {
+        return;
+      }
+      LOG.debug("resourceOffers: Currently have " + _offers.size() + " offers buffered" +
+          (_offers.size() > 0 ? (":" + offerMapToString(_offers)) : ""));
+      for (Protos.Offer offer : offers) {
+        if (isHostAccepted(offer.getHostname())) {
+          LOG.debug("resourceOffers: Recording offer from host: " +
+              offer.getHostname() + ", offerId: " + offer.getId().getValue());
+          _offers.put(offer.getId(), offer);
+        } else {
+          LOG.debug("resourceOffers: Declining offer from host: " +
+              offer.getHostname() + ", offerId: " + offer.getId().getValue());
+          driver.declineOffer(offer.getId());
+        }
+      }
+      LOG.debug("resourceOffers: After processing offers, now have " +
+          _offers.size() + " offers buffered:" + offerMapToString(_offers));
+    }
+  }
+
+  public void offerRescinded(OfferID id) {
+    synchronized (_offersLock) {
+      _offers.remove(id);
+    }
+  }
+
+  public void taskLost(final TaskID taskId) {
```

Comment:
Yeah - but you're not doing task reconciliation. So, if you're dependent on task statuses working at all, things wont work well.